HULKs/hulk
# Simulator Generator

## Why? What?

Node execution glue code is automatically generated by our framework for the normal hulk_* binaries.
However, the behavior simulator always used a once copied and since hand maintained approach, which has caused a lot of issues.
This PR removes the manually maintained cycler and replaces it with a generated one.

Other notable changes:
- Moved the behavior simulator from tools to `crates/hulk_behavior_simulator`. Thus it can be used via `pepsi run --target behavior_simulator -- ...`.
- Previously scenarios were changing the `filtered_game_controller_state` directly, but the new cycler executies the `game_controller_state_filter` node which arguably is part of our behavior. The scenarios were adjusted to change the `game_controller_state`.
- Added some basic documentation

Based on #1002 

Fixes #

## ToDo / Known Issues


## Ideas for Next Iterations (Not This PR)

- RIIR, Lua bad tbh.
- Provide game controller like API to change the game controller state instead of having to manually in the state herumhuren.

## How to Test

Common pain points with the behavior simulator such as adding/removing/renaming cycle context fields of nodes executed in the behavior simulator should now no longer occur.
Note that when adding an input that isn't currently produced in the behavior simulator, you either have to include the node that produces it in the build.rs or add the output to the fake data node.
But you no longer have to change the cycle context instantiation yourself.
